File: quote/Projects/PluginModelIO/TemplateLoader.vb
Imports System.Data.SqlClient
Imports System.Reflection
Imports System.Data.OleDb
Imports System.Transactions

Imports DB.QuoteDataBaseTableAdapters
Imports DB.QuoteDataBase
Imports DB

Imports Model.Template
Imports Model
Imports System.Windows.Forms

Public Class TemplateLoader

    Public Function Load(ByVal id As Long) As Header

        System.Windows.Forms.Cursor.Current = Cursors.WaitCursor

        Dim adaptor As New QuoteDataBaseTableAdapters._QuoteTableAdapter
        Dim table As New QuoteDataBase._QuoteDataTable
        Dim q As New Header()

        adaptor.FillByByQuoteID(table, id)
        If table.Rows.Count > 0 Then
            Dim row As QuoteDataBase._QuoteRow = table.Rows(0)
            q = New Header(row.id)

            Dim rfq As String = ""
            If Not row.IsRequestForQuoteNumberNull Then
                rfq = row.RequestForQuoteNumber
            End If
            Dim part As String = ""
            If Not row.IsPartNumberNull Then
                part = row.PartNumber
            End If
            Dim Initials As String = ""
            If Not row.IsInitialsNull Then
                Initials = row.Initials
            End If
            Dim createdDate As DateTime
            If Not row.IsCreatedDateNull Then
                createdDate = row.CreatedDate
            End If
            Dim lastModDate As DateTime
            If Not row.IsLastModifedDateNull Then
                lastModDate = row.LastModifedDate
            End If

            Dim customerObj As Customer = LookupCustomer(row)

            q.PrimaryProperties.CommonCustomer = customerObj
            q.PrimaryProperties.CommonPartNumber = part
            q.PrimaryProperties.CommonRequestForQuoteNumber = rfq
            q.PrimaryProperties.CommonCreatedDate = createdDate
            q.PrimaryProperties.CommonLastModified = lastModDate
            q.PrimaryProperties.CommonInitials = Initials

            CommonLoader.LoadComputationProperties(id, q.ComputationProperties)
            CommonLoader.LoadOtherProperties(id, q.OtherProperties)
            CommonLoader.LoadNoteProperties(id, q.NoteProperties)
            LoadComponents(q)
        End If

        q.ComputationProperties.ClearDirty()
        q.OtherProperties.ClearDirty()
        q.PrimaryProperties.ClearDirty()
        q.NoteProperties.ClearDirty()
        q.ClearDirty()

        System.Windows.Forms.Cursor.Current = Cursors.Default

        Return q
    End Function

    Public Function LookupCustomer(row As QuoteDataBase._QuoteRow) As Customer

        Dim customer As String = ""
        If Not row.IsCustomerNameNull Then
            customer = row.CustomerName
        End If
        Dim customerID As Integer

        If row.IsCustomerIDNull Then
            If Not row.IsCustomerNameNull Then
                Dim temp As Customer
                temp = Model.Template.Customer.GetByName(row.CustomerName)
                If (Not temp Is Nothing) Then
                    customerID = temp.ID
                End If
            End If
        Else
            customerID = row.CustomerID
        End If

        Dim customerObj As New Customer
        customerObj.SetName(customer)
        customerObj.SetID(customerID)

        Return customerObj
    End Function


    Public Shared Sub LoadComponents(ByVal q As Model.Common.Header)

        Dim adaptor As New _QuoteDetailTableAdapter
        Dim id As Integer = q.PrimaryProperties.CommonID
        Dim table As _QuoteDetailDataTable = adaptor.GetDataByQuoteID(id)
        For Each row As _QuoteDetailRow In table.Rows

            Dim temp As New TempObj
            CommonLoader.LoadProperties(id, row.id, temp)

            Dim product As New Model.Product( _
              row.ProductCode, _
              temp.Gage, _
              0, _
              0, _
              row.IsWire, _
              "", _
              0,
              "", _
              0, _
              0)
            Dim detail As Model.Common.Detail = q.NewDetail(product)
            detail.Qty = row.Qty
            With detail
                .IsWire = row.IsWire
                If Not row.IsSourceIDNull Then .SourceID = row.SourceID
                .SequenceNumber = row.SequenceNumber
            End With
            CommonLoader.LoadProperties(id, row.id, detail.QuoteDetailProperties)
        Next
    End Sub

    Public Class TempObj
        Public Property UnitOfMeasure As String
        Public Property Gage As String = ""
    End Class

End Class

File: trunk/quote/Projects/PluginModelIO/TemplateLoader.cs
namespace Model.IO
{
    using System;
    using System.Linq;
    using System.Windows.Forms;
    
    using DB;
    using DB.QuoteDataBaseTableAdapters;
    
    using Model.Template;

    public static class TemplateLoader
    {
        public static Header Load(int id)
        {
            System.Windows.Forms.Cursor.Current = Cursors.WaitCursor;

            DB.QuoteDataBaseTableAdapters._QuoteTableAdapter adaptor = 
                new DB.QuoteDataBaseTableAdapters._QuoteTableAdapter();
            DB.QuoteDataBase._QuoteDataTable table = new DB.QuoteDataBase._QuoteDataTable();
            Header q = new Header();

            adaptor.FillByByQuoteID(table, id);
            if (table.Rows.Count > 0) 
            {
                DB.QuoteDataBase._QuoteRow row = table.Rows[0] as DB.QuoteDataBase._QuoteRow;
                q = new Header((int)row.id);

                string rfq = string.Empty;
                if (!row.IsRequestForQuoteNumberNull()) 
                {
                    rfq = row.RequestForQuoteNumber;
                }
                
                string part = string.Empty;
                if (!row.IsPartNumberNull()) 
                {
                    part = row.PartNumber;
                }
                
                string initials = string.Empty;
                if (!row.IsInitialsNull()) 
                {
                    initials = row.Initials;
                }
                
                DateTime createdDate = default(DateTime);
                if (!row.IsCreatedDateNull()) 
                {
                    createdDate = row.CreatedDate;
                }
                
                DateTime lastModDate = default(DateTime);
                if (!row.IsLastModifedDateNull()) 
                {
                    lastModDate = row.LastModifedDate;
                }

                Customer customerObj = LookupCustomer(row);

                q.PrimaryProperties.CommonCustomer = customerObj;
                q.PrimaryProperties.CommonPartNumber = part;
                q.PrimaryProperties.CommonRequestForQuoteNumber = rfq;
                q.PrimaryProperties.CommonCreatedDate = createdDate;
                q.PrimaryProperties.CommonLastModified = lastModDate;
                q.PrimaryProperties.CommonInitials = initials;

                CommonLoader.LoadComputationProperties(id, q.ComputationProperties);
                CommonLoader.LoadOtherProperties(id, q.OtherProperties);
                CommonLoader.LoadNoteProperties(id, q.NoteProperties);
                LoadComponents(q);
            }

            q.IsDirty = false;
            
            System.Windows.Forms.Cursor.Current = Cursors.Default;

            return q;
        }

        public static Customer LookupCustomer(DB.QuoteDataBase._QuoteRow row)
        {
            string customer = string.Empty;
            if (!row.IsCustomerNameNull()) 
            {
                customer = row.CustomerName;
            }
            
            int customerID = 0;

            if (row.IsCustomerIDNull()) 
            {
                if (!row.IsCustomerNameNull()) 
                {
                    Customer temp = null;
                    
                    temp = Model.IO.Misc.CustomerDB.GetByName(row.CustomerName);
                    
                    if (temp != null)
                    {
                        customerID = temp.Id;
                    }
                }
            } 
            else
            {
                customerID = (int)row.CustomerID;
            }

            Customer customerObj = new Customer();
            customerObj.SetName(customer);
            customerObj.SetId(customerID);

            return customerObj;
        }

        public static void LoadComponents(Model.Common.Header header)
        {
            _QuoteDetailTableAdapter adaptor = new _QuoteDetailTableAdapter();
            
            int id = header.PrimaryProperties.CommonId;
            
            DB.QuoteDataBase._QuoteDetailDataTable table = adaptor.GetDataByQuoteID(id);

            foreach (DB.QuoteDataBase._QuoteDetailRow row in table.Rows) 
            {
                TemplateLoaderTempObj temp = new TemplateLoaderTempObj();
                CommonLoader.LoadProperties(id, (int)row.id, temp);

                ProductBuildData data = new ProductBuildData();
                data.Code = row.ProductCode;
                data.Gage = temp.Gage;
                data.IsWire = row.IsWire;
                data.Vendor = string.Empty;
                data.Description = string.Empty;
                Model.Product product = new Model.Product(data);

                Model.Common.Detail detail = header.NewDetail(product);
                detail.Qty = row.Qty;
                
                if (!row.IsSourceIDNull())
                {
                    detail.SourceId = row.SourceID;
                }
                
                detail.SequenceNumber = row.SequenceNumber;
                header.NextSequenceNumber = detail.SequenceNumber + 1;
                
                CommonLoader.LoadProperties(id, (int)row.id, detail.QuoteDetailProperties);
            }
        }
    }
}
